### PR TITLE
Values

### PR DESCRIPTION
**Pull request by Montgomeryyyy**
This pull request introduces values, value normalisation, and prioritised truncation, which can be useful when using values.  #36 

### Data Normalization and Value Handling:
* [`corebehrt/classes/normalizer.py`](diffhunk://#diff-44927c449792ac8a26c1105792b735b479adfcff48f0826c83df91d6eea36552R1-R71): Introduced `ValuesNormalizer` class for normalizing values in data frames, including a `min_max_normalize_results` method.
* [`corebehrt/functional/values.py`](diffhunk://#diff-03e6d069016c0e651068cf416677951a0d59d8f517bd406df8fd71873c83e854R1-R46): Added methods `add_binned_values` and `add_quantile_values` to handle value binning and quantile values in data frames.
* [`corebehrt/functional/normalize.py`](diffhunk://#diff-1898426388e489a1dec04a0249b3276816607010bf73ebf562777be790f062f2R1-R5): Added a `min_max_normalize` function for min-max normalization.

### Enhancements to `FormattedDataLoader`:
* [`corebehrt/classes/loader.py`](diffhunk://#diff-012462f80bf201c72551037e98cd75717df52cd0b05687aa8efa1762f4fd5b39L1-R13): Updated `FormattedDataLoader` to include new parameters for value handling (`include_values`, `value_type`, `normalise_func`) and added `_apply_values` method to apply these values. [[1]](diffhunk://#diff-012462f80bf201c72551037e98cd75717df52cd0b05687aa8efa1762f4fd5b39L1-R13) [[2]](diffhunk://#diff-012462f80bf201c72551037e98cd75717df52cd0b05687aa8efa1762f4fd5b39L31-R57) [[3]](diffhunk://#diff-012462f80bf201c72551037e98cd75717df52cd0b05687aa8efa1762f4fd5b39R75-R86)

### Priority-based Truncation:
* [`corebehrt/data/prepare_data.py`](diffhunk://#diff-f039261bf5bac76d90e24097dece7a6ddbbc7c573b59bcbe0f55da53535aaca3R163-R168): Updated data preparation methods to include priority-based truncation using `prioritized_truncate_patient`. [[1]](diffhunk://#diff-f039261bf5bac76d90e24097dece7a6ddbbc7c573b59bcbe0f55da53535aaca3R163-R168) [[2]](diffhunk://#diff-f039261bf5bac76d90e24097dece7a6ddbbc7c573b59bcbe0f55da53535aaca3R236-R242)
* [`corebehrt/functional/utils.py`](diffhunk://#diff-27a9ce472c023ccc54798ce4026301c0a261a3885a3e34ed3cb2c7254b184f49R169-R228): Added `prioritized_truncate_patient` method and updated `truncate_data` to accept additional keyword arguments for truncation. [[1]](diffhunk://#diff-27a9ce472c023ccc54798ce4026301c0a261a3885a3e34ed3cb2c7254b184f49R169-R228) [[2]](diffhunk://#diff-27a9ce472c023ccc54798ce4026301c0a261a3885a3e34ed3cb2c7254b184f49R272-R300)

### Configuration and Utility Updates:
* [`corebehrt/functional/utils.py`](diffhunk://#diff-27a9ce472c023ccc54798ce4026301c0a261a3885a3e34ed3cb2c7254b184f49R272-R300): Added `init_function` utility to dynamically initialize functions or class methods from a path string.

### Miscellaneous:
* [`corebehrt/classes/features.py`](diffhunk://#diff-b722d8130331f234335bdc116f71a2d23318d35aa06a0a0d57b398281d7716dfR12): Added `assign_index_and_order` function to ensure correct ordering in data frames. [[1]](diffhunk://#diff-b722d8130331f234335bdc116f71a2d23318d35aa06a0a0d57b398281d7716dfR12) [[2]](diffhunk://#diff-b722d8130331f234335bdc116f71a2d23318d35aa06a0a0d57b398281d7716dfR63-R64)

### CODE REVIEW

File: corebehrt/classes/loader.py [hunk 1]
```diff
@@ -28,14 +32,29 @@ class FormattedDataLoader:
             Returns a tuple containing the concepts DataFrame and the patients_info DataFrame.
     """
 
-    def __init__(self, folder: str, concept_types: List[str]):
+    def __init__(
+        self,
+        folder: str,
+        concept_types: List[str],
+        include_values: List[str] = [],
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/classes/loader.py:39]: added in 65ec45b

File: corebehrt/functional/creators.py [hunk 2]
```diff
@@ -139,9 +139,34 @@ def assign_segments_to_death(df: dd.DataFrame) -> dd.DataFrame:
...
 
 
+def assign_index_and_order(df: dd.DataFrame) -> dd.DataFrame:
+    """
+    Assign 'index' and 'order' columns to ensure correct ordering.
+    Initialises 'index' and 'order' columns with 0 to prioritise unassigned rows in sorting.
+    Parameters:
+        df: dd.DataFrame with 'PID' column.
+    Returns:
+        df with 'index' and 'order' columns.
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/functional/creators.py:114]: added in 65ec45b

File: corebehrt/functional/utils.py [hunk 3]
```diff
@@ -172,24 +166,66 @@ def truncate_patient(
     return truncated_patient
 
 
+def prioritized_truncate_patient(
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/functional/utils.py:216]: added in b5089fc

File: corebehrt/functional/utils.py [hunk 4]
```diff
@@ -172,24 +166,66 @@ def truncate_patient(
...
+    patient: pd.DataFrame,
+    background_length: int,
+    max_len: int,
+    sep_token: str,
+    low_priority_prefixes: List[str],
+    vocabulary: dict,
+    unit: bool = False,
+) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/functional/utils.py:224]: added in b5089fc

File: corebehrt/functional/values.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+from typing import Optional, Callable
+import pandas as pd
+
+
+def add_binned_values(
+    concepts: dd.DataFrame,
+    normalise: Optional[Callable[[pd.Series], pd.Series]] = None,
+) -> dd.DataFrame:
+    """
+    Includes binning values for the concepts DataFrame. Only includes numeric values.
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/functional/values.py]: improved in 65ec45b

File: corebehrt/functional/values.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    Expects 'RESULT' and 'CONCEPT' column to be present.
+    Optionally normalises the 'RESULT' column using the provided normalise function.
+    """
+    concepts["RESULT"] = dd.to_numeric(concepts["RESULT"], errors="coerce")
+    concepts["index"] = concepts.index + 1
+
+    values = concepts.dropna(subset=["RESULT"])
+    if normalise is not None and callable(normalise):
+        values = normalise(values)
+    values["RESULT"] = (values["RESULT"] * 100).astype(int)
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/functional/values.py]: added in fad59a7, now a arguments dictionary is given as input

File: corebehrt/functional/values.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    concepts["order"] = 0
+    values["order"] = 1
+    concatted_df = dd.concat([concepts, values])
+    return concatted_df.drop(["RESULT"], axis=1)
+
+
+def add_quantile_values(
+    concepts: dd.DataFrame,
+) -> dd.DataFrame:
+    """
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/functional/values.py]: added in 65ec45b

File: corebehrt/classes/normalizer.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    @staticmethod
+    def min_max_normalize_results(values: dd.DataFrame) -> dd.DataFrame:
+        """
+        Normalises the 'RESULT' column of the given DataFrame using min-max normalisation.
+        If the count of a CONCEPT is less than 3, the RESULT is set to -1.
+        Expects 'RESULT' and 'CONCEPT' column to be present.
+        """
+        # Compute min, max, and count for each CONCEPT
+        grouped = values.groupby("CONCEPT")["RESULT", "TIMESTAMP"]
+        min_values = (
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/classes/normalizer.py]: improved in fad59a7

File: corebehrt/classes/features.py [hunk 9]
```diff
@@ -59,6 +60,8 @@ def __call__(
         death = create_death(patients_info)
         features = dd.concat([concepts, background, death])
 
+        features = assign_index_and_order(features)
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/classes/features.py]: I have now moved the sorting to its own function in 4dbc288. For the index and order, they need to be filled out before we remove rows with NaNs, otherwise background tokens, etc. will be removed. I now moved it to right before removing the NaNs

File: corebehrt/functional/utils.py [hunk 10]
```diff
@@ -233,3 +269,32 @@ def check_required_columns(
     if not required_columns.issubset(set(df.columns)):
         missing_columns = required_columns - set(df.columns)
         raise ValueError(f"Missing columns in {type_}: {missing_columns}")
+
+
+def init_function(func_path: str):
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/functional/utils.py]: added in b5089fc

File: corebehrt/main/create_data.py [hunk 11]
```diff
@@ -150,7 +151,23 @@ def create_and_save_features(excluder: Excluder, cfg) -> None:
     """
     save_path = join(cfg.output_dir, cfg.paths.save_features_dir_name)
     concepts, patients_info = FormattedDataLoader(
-        cfg.loader.data_dir, cfg.loader.concept_types
+        cfg.loader.data_dir,
```
[the remark this answers is not in the document]
Comment (Montgomeryyyy) [reply corebehrt/main/create_data.py]: now moved to its own class in 518172f

**Comment by Montgomeryyyy** [comment]
I think I've addressed all of the comments now

**Review by kirilklein — APPROVED** [review]

**Comment by kirilklein** [comment]
Very good! 
Maybe we can also ask Mikkel or Stephan to have a look before merging :)

File: corebehrt/classes/normalizer.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    A class to load normalise values in data frames.
+    Expects 'RESULT' and 'concept' column to be present.
+    """
+
+    @staticmethod
+    def apply_normalization(row, normalization_func, **kwargs):
+        """
+        Applies the given normalisation function if count >= min_count, else sets RESULT to -1.
+        """
+        min_count = kwargs.pop("min_count", 3)
```
Comment (mikkelfo) [inline corebehrt/classes/normalizer.py:16]: Shouldn't we just include `min_count=3` as a keyword argument, instead of hiding the dependency in `**kwargs`?

File: corebehrt/classes/values.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    >>> features = feature_creator(patients_info, concepts)
+    """
+
+    def __init__(
+        self,
+        value_type: str,
+        value_type_kwargs: dict = None,
+        normalize_args: dict = None,
+    ):
+        self.value_type = value_type  # values.get("value_type", None)
```
Comment (mikkelfo) [inline corebehrt/classes/values.py:28]: Probably also just cleanup the comments

File: corebehrt/classes/values.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+            )
+
+    def __call__(
+        self,
+        concepts: dd.DataFrame,
+    ) -> dd.DataFrame:
+        if self.value_type == "binned":
+            return self.add_binned_values(
+                concepts,
+                self.value_type_kwargs["multiplication_factor"],
```
Comment (mikkelfo) [inline corebehrt/classes/values.py:45]: This will fail if self.value_type_kwargs is `None`, as is the default value? So either we add a `.get` or we remove the default `None` in the `__init__`

File: corebehrt/data/prepare_data.py [hunk 15]
```diff
@@ -219,7 +220,7 @@ def _prepare_mlm_features(self, predefined_splits, val_ratio=0.2) -> Data:
 
         # 5. Truncation
         logger.info(f"Truncating data to {data_cfg.truncation_len} tokens")
-        data = truncate_data(data, data_cfg.truncation_len, vocab, truncate_patient)
+        data = self._truncate_data(data, vocab, data_cfg)
```
Comment (mikkelfo) [inline corebehrt/data/prepare_data.py:223]: Don't exactly remember what we decided here, but I believe we shouldn't pass `data_cfg` around and instead only pass in the relevant elements of `data_cfg` into the function (In this case `priority_truncation` and `truncation_len`

File: corebehrt/functional/normalize.py [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import numpy as np
+
+# def min_max_normalize(value, min_value, max_value) -> float:
```
Comment (mikkelfo) [inline corebehrt/functional/normalize.py:3]: Delete if not relevant anymore

File: corebehrt/main/create_data.py [hunk 17]
```diff
@@ -117,22 +118,29 @@ def create_and_save_features(excluder: Excluder, cfg) -> None:
     Returns a list of lists of pids for each batch
     """
     concepts, patients_info = FormattedDataLoader(
-        cfg.paths.data, cfg.loader.concept_types
+        cfg.paths.data,
+        cfg.loader.concept_types,
+        include_values=(
+            cfg.loader.include_values
+            if hasattr(cfg.loader, "include_values") and cfg.loader.include_values
```
Comment (mikkelfo) [inline corebehrt/main/create_data.py:125]: Can probably just be `getattr(cfg.loader, "include_values", False)` for simplicity

**Review by mikkelfo — CHANGES_REQUESTED** [review]
Sorry for being slow for the review! There's just some minor changes I'd like, since we'll have to do them eventually, so might as well do it now. Sorry!

File: corebehrt/functional/utils.py [hunk 18]
```diff
@@ -172,24 +165,111 @@ def truncate_patient(
...
+def _drop_non_priority_tokens(
+    patient: pd.DataFrame,
+    non_priority_tokens: set,
+    truncation_length: int,
+    background_length: int,
+) -> pd.DataFrame:
+    """
+    Drop non-priority tokens from patient, keeping truncation_length - background_length tokens.
+    """
+    patient["non_priority"] = patient["concept"].isin(non_priority_tokens)
```
Comment (DrLor) [inline corebehrt/functional/utils.py:188]: I think it also modifies the patient df. Maybe just save the `non_priority` Series into a temporary var instead.

Comment (DrLor) [reply corebehrt/functional/utils.py:188]: Or, as `non_priority` seems to be used in the function below, maybe have a different function for computing it and adding it. Currently `_filter_invalid_positions` will fail if `_drop_non_priority_tokens` has not been called beforehand.

File: corebehrt/main/create_data.py (hunk 17 above)
Comment (StephanLorenzen) [reply corebehrt/main/create_data.py:125]: Or `cfg.loader.get("include_values")` would work as well?

File: tests/data/raw_with_values/concept.labs.csv
Comment (StephanLorenzen) [inline tests/data/raw_with_values/concept.labs.csv:1]: Seems not to be used.

**Review by StephanLorenzen — CHANGES_REQUESTED** [review]
Very nice!
I agreee with Mikkels comments and also put some comments, mostly related to side-effects in functionals and some unused test data.

**Review by kirilklein — APPROVED** [review]
Approved

**Review by StephanLorenzen — APPROVED** [review]

**Review by mikkelfo — APPROVED** [review]
